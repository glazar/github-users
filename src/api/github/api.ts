import { GithubUser } from './api';
export interface GithubUser {
  id: number;
  login: string;
  avatar_url: string;
  html_url: string;
}

export const fetchUserList = (since?: number): Promise<GithubUser[]> => {
  let url = 'https://api.github.com/users';

  if (since) {
    url += `?since=${since}`;
  }

  return fetch(url).then(response => response.json<GithubUser[]>());
};

export const fetchUser = (login: string): Promise<GithubUser> => {
  const url = `https://api.github.com/users/${login}`;

  return fetch(url).then(response => {
    if (response.status === 404) {
      throw response;
    } else {
      return response.json<GithubUser>();
    }
  });
};
