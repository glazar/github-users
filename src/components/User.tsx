import * as React from 'react';
import { SFC } from 'react';
import { Alert, Button, PageHeader } from 'react-bootstrap';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import { RouteComponentProps } from 'react-router-dom';
import { ThunkDispatch } from 'redux-thunk';

import { clearUser, ClearUserAction, getUser, GetUserActions } from '../actions/userActions';
import { GithubUser } from '../api/github/api';
import { State, UserState } from '../store';
import { Loader } from './Spinner';
import { UserThumbnail } from './UserThumbnail';

interface UserStateProps {
  user?: GithubUser;
  isFetching?: boolean;
}

interface UserDispatchProps {
  onLoadUser?: () => void;
  onClearUser?: () => void;
}

type UserOwnProps = RouteComponentProps<{ login: string }, {}>;

const mapStateToProps = (state: State, ownProps: UserOwnProps): UserStateProps => {
  return {
    user: state.userState.user
      ? state.userState.user
      : state.userListState.userList
        ? state.userListState.userList.find(user => user.login === ownProps.match.params.login)
        : state.userState.user,
    isFetching: state.userState.isFetching
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<State, any, GetUserActions | ClearUserAction>,
  ownProps: UserOwnProps
): UserDispatchProps => {
  return {
    onLoadUser() {
      dispatch(getUser(ownProps.match.params.login));
    },
    onClearUser() {
      dispatch(clearUser());
    }
  };
};

type UserProps = UserState & UserDispatchProps;

export class User extends React.PureComponent<UserProps, {}> {
  public componentDidMount(): void {
    if (!this.props.user) {
      this.handleLoadUser();
    }
  }

  public componentWillUnmount(): void {
    this.props.onClearUser!();
  }

  public render(): JSX.Element {
    const { user, isFetching } = this.props;

    return (
      <div>
        <LinkContainer to="/">
          <Button bsStyle="primary" className="m-2" bsSize="large">
            User List
          </Button>
        </LinkContainer>

        <PageHeader>User Details</PageHeader>

        {user && !isFetching && <UserCard user={user} />}

        {!user && !isFetching && <Alert bsStyle="warning">No Data Available</Alert>}

        {isFetching && <Loader />}

        {!isFetching && (
          <Button bsStyle="primary" className="m-2" bsSize="large" onClick={this.handleLoadUser}>
            Reload User Details
          </Button>
        )}
      </div>
    );
  }

  private handleLoadUser = (): void => {
    this.props.onLoadUser!();
  };
}

export const UserCard: SFC<{ user: GithubUser }> = ({ user }) => (
  <UserThumbnail user={user}>
    <p>
      ID: <span className="badge badge-info pl-1 pr-1">{user.id}</span>
    </p>
    <p>
      URL:
      <a className="nav-link d-inline pl-1 pr-1" target="_blank" href={user.html_url}>
        <span className="badge badge-info"> {user.html_url}</span>
      </a>
    </p>
  </UserThumbnail>
);

export const ConnectedUser = connect(
  mapStateToProps,
  mapDispatchToProps
)(User);

export default ConnectedUser;
