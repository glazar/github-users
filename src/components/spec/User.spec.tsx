import { shallow } from 'enzyme';
import * as React from 'react';

import { GithubUser } from '../../api/github/api';
import { User, UserCard } from '../User';

it('UserCard component renders user data properly', () => {
  const user: GithubUser = {
    id: 1,
    avatar_url: 'https://avatars3.githubusercontent.com/u/4364303?v=4',
    login: 'glazar',
    html_url: 'https://github.com/glazar'
  };
  const wrapper = shallow(<UserCard user={user} />);
  expect(wrapper).toMatchSnapshot();
});

it('User component should call to load user if no user provided', () => {
  const mockHandleLoadUser = jest.fn();
  shallow(<User onLoadUser={mockHandleLoadUser} />);
  expect(mockHandleLoadUser).toHaveBeenCalledTimes(1);
});

it('User component should not call to load user if user provided', () => {
  const mockHandleLoadUser = jest.fn();
  const user: GithubUser = {
    id: 1,
    avatar_url: 'https://avatars3.githubusercontent.com/u/4364303?v=4',
    login: 'glazar',
    html_url: 'https://github.com/glazar'
  };
  shallow(<User user={user} onLoadUser={mockHandleLoadUser} />);
  expect(mockHandleLoadUser).not.toHaveBeenCalled();
});
