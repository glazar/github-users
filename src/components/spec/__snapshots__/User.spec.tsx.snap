// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard component renders user data properly 1`] = `
<Component
  user={
    Object {
      "avatar_url": "https://avatars3.githubusercontent.com/u/4364303?v=4",
      "html_url": "https://github.com/glazar",
      "id": 1,
      "login": "glazar",
    }
  }
>
  <p>
    ID: 
    <span
      className="badge badge-info pl-1 pr-1"
    >
      1
    </span>
  </p>
  <p>
    URL:
    <a
      className="nav-link d-inline pl-1 pr-1"
      href="https://github.com/glazar"
      target="_blank"
    >
      <span
        className="badge badge-info"
      >
         
        https://github.com/glazar
      </span>
    </a>
  </p>
</Component>
`;
